src/Roslyn.Diagnostics.Analyzers/Core/TemporaryArrayAsRefAnalyzer.cs

```diff
@@ -0,0 +1,74 @@
+    public class TemporaryArrayAsRefAnalyzer : DiagnosticAnalyzer
+    {
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(RoslynDiagnosticsAnalyzersResources.TemporaryArrayAsRefTitle), RoslynDiagnosticsAnalyzersResources.ResourceManager, typeof(RoslynDiagnosticsAnalyzersResources));
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(RoslynDiagnosticsAnalyzersResources.TemporaryArrayAsRefMessage), RoslynDiagnosticsAnalyzersResources.ResourceManager, typeof(RoslynDiagnosticsAnalyzersResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(RoslynDiagnosticsAnalyzersResources.TemporaryArrayAsRefDescription), RoslynDiagnosticsAnalyzersResources.ResourceManager, typeof(RoslynDiagnosticsAnalyzersResources));
+
+        internal static DiagnosticDescriptor Rule = new(
+            RoslynDiagnosticIds.TemporaryArrayAsRefRuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.RoslynDiagnosticsReliability,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            description: s_localizableDescription,
+            helpLinkUri: null,
+            customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                var temporaryArrayExtensions = context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftCodeAnalysisSharedCollectionsTemporaryArrayExtensions);
+                if (temporaryArrayExtensions is null)
+                    return;
+
+                var temporaryArrayAsRef = (IMethodSymbol)temporaryArrayExtensions.GetMembers("AsRef").SingleOrDefault();
+                if (temporaryArrayAsRef is null)
+                    return;
+
+                context.RegisterOperationAction(context => AnalyzeInvocation(context, temporaryArrayAsRef), OperationKind.Invocation);
+            });
+        }
+
+        private static void AnalyzeInvocation(OperationAnalysisContext context, IMethodSymbol temporaryArrayAsRef)
+        {
+            var invocation = (IInvocationOperation)context.Operation;
+            var targetMethod = invocation.TargetMethod.ReducedFrom ?? invocation.TargetMethod;
+            if (!Equals(targetMethod.OriginalDefinition, temporaryArrayAsRef))
+                return;
+
+            var instance = invocation.Instance ?? invocation.Arguments.FirstOrDefault()?.Value;
```

Comment:
Use the extension method `IInvocationOperation.GetReceiverType`? If that doesn't work, seems to be a good candidate to add a parallel extension method that matches your scenario.

Reply:
I'm not sure that helps here (don't need the type).